### PR TITLE
[WIP] Stub out InMemoryCASDatabase.

### PR DESCRIPTION
**Pull request by ddunbar**
* [ ] Depends on #665

### CODE REVIEW

**Comment by ddunbar** [comment]
This PR isn't ready, just here to show where I am going with #664 and #665...

I am quite new to `std::future`, among other things...

File: include/llbuild/CAS/CASDatabase.h [hunk 1]
```diff
@@ -50,8 +51,18 @@ class CASDatabase {
...
   /// Write the given object into the database.
-  virtual auto put(std::unique_ptr<CASObject> object) ->
+  virtual auto putObject(std::unique_ptr<CASObject> object) ->
     std::future<llvm::ErrorOr<DataID>> = 0;
+
+  /// Write an object into the database.
+  inline auto put(std::unique_ptr<CASObject> object)
+    -> std::future<llvm::ErrorOr<DataID>>
+  {
+    return putObject(std::move(object));
```
Comment (dmbryson) [inline include/llbuild/CAS/CASDatabase.h:61]: Ah ha, I see here why the unique_ptr makes sense... for in memory.  🤔 Saves a copy going in, but we still have to copy it coming out.  Probably too early to say for certain, but seems like there is potential a lot of copies happening here.

File: include/llbuild/CAS/CASDatabase.h [hunk 2]
```diff
@@ -50,8 +51,18 @@ class CASDatabase {
     std::future<llvm::ErrorOr<std::unique_ptr<CASObject>>> = 0;
 
   /// Write the given object into the database.
-  virtual auto put(std::unique_ptr<CASObject> object) ->
+  virtual auto putObject(std::unique_ptr<CASObject> object) ->
```
Comment (dmbryson) [inline include/llbuild/CAS/CASDatabase.h:54]: Why `put` and `putObject`?

File: lib/CAS/InMemoryCASDatabase.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  return p.get_future();
+}
+
+auto InMemoryCASDatabase::putObject(std::unique_ptr<CASObject> object) ->
+  std::future<llvm::ErrorOr<DataID>>
+{
+  std::unique_lock<std::mutex> lock(mutex);
+
+  // FIXME: Actually hash the object.
+  DataID hash("somehash");
```
Comment (dmbryson) [inline lib/CAS/InMemoryCASDatabase.cpp:55]: I like this hash algorithm.  😉

File: unittests/CAS/InMemoryCASDatabaseTests.cpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+#include "gtest/gtest.h"
+
+using namespace llbuild;
+using namespace llbuild::CAS;
+
+TEST(InMemoryCASDatabaseTests, basic) {
+  InMemoryCASDatabase db;
+
+  // Put an trivial object.
+  auto id = db.put("X").get().get();
```
Comment (dmbryson) [inline unittests/CAS/InMemoryCASDatabaseTests.cpp:24]: `.get().get()`...    sadness.   Makes sense, but not particularly friendly/ergonomic.

Don't know llvm::ErrorOr... what happens if it actually contains an error?

**Comment by shahmishal** [comment]
The Swift project moved the default branch to `main` and deleted `master` branch, so GitHub automatically closed the PR. Please re-create pull request with `main` branch.

More detail about the branch update - https://forums.swift.org/t/updating-branch-names/40412

**Pull request closed without merging**